oak_functions/wasm/src/lib.rs

```diff
@@ -36,388 +36,470 @@ use oak_functions_abi::{
-            )),
-            _ => {
-                // Here https://paritytech.github.io/wasmi/wasmi/trait.Externals.html#examples panics with
-                //  panic!("Unimplemented function at {}.", index)
-                // We prefer not to panic, and trap in an unreachable state instead.
-                Err(wasmi::Trap::new(wasmi::TrapKind::Unreachable))
-            }
-        }
+    fn get_response_bytes(&self) -> Vec<u8> {
+        let user_state = self.store.data();
+        user_state.response_bytes.clone()
     }
-}
 
-impl<L> wasmi::ModuleImportResolver for WasmState<L>
-where
-    L: OakLogger,
-{
-    fn resolve_func(
-        &self,
-        field_name: &str,
-        signature: &wasmi::Signature,
-    ) -> Result<wasmi::FuncRef, wasmi::Error> {
-        // Look for the function (i.e., `field_name`) in the statically registered functions.
-        let (index, expected_signature) =
-            oak_functions_resolve_func(field_name).ok_or_else(|| {
-                wasmi::Error::Instantiation(format!("export {} not found", field_name))
-            })?;
+    // Validates whether a given address range (inclusive) falls within the currently allocated
+    // range of guest memory.
+    /* TODO(mschett): Check if we still need this.
 
-        if signature == &expected_signature {
-            Ok(wasmi::FuncInstance::alloc_host(expected_signature, index))
+    fn validate_range(&self, addr: AbiPointer, offset: AbiPointerOffset) -> Result<(), OakStatus> {
+        let memory = self
+            .instance
+            .get_export(&self.store, "memory")
+            // TODO(mschett): Fix unwrap.
+            .unwrap()
+            .into_memory()
+            .expect("WasmState memory not attached!?");
+
+        // TODO(mschett): Check if there is a better way to check the memory size.
+        let memory_size: wasmi::core::memory_units::Bytes =
+            wasmi::core::memory_units::Pages::from(memory.current_pages(&self.store)).into();
+        // Check whether the end address is below or equal to the size of the guest memory.
+        if wasmi::core::memory_units::Bytes((addr as usize) + (offset as usize)) <= memory_size {
+            Ok(())
         } else {
-            Err(wasmi::Error::Instantiation(format!(
-                "export `{}` doesn't match expected signature; got: {:?}, expected: {:?}",
-                field_name, signature, expected_signature
-            )))
+            Err(OakStatus::ErrInvalidArgs)
         }
     }
+     */
+}
+
+/// Reads the buffer starting at address `buf_ptr` with length `buf_len` from the Wasm memory.
+pub fn read_buffer(
+    ctx: &mut impl AsContext,
+    memory: &mut wasmi::Memory,
+    buf_ptr: AbiPointer,
+    buf_len: AbiPointerOffset,
+) -> Result<Vec<u8>, OakStatus> {
+    let mut target = alloc::vec![0; buf_len as usize];
+
+    let buf_ptr = usize::try_from(buf_ptr)
+        .expect("failed to convert AbiPointer to usize as required by wasmi API");
```

Comment:
I think `let buf_ptr: usize = buf_ptr.try_into()` or `let buf_ptr = buf_ptr.try_into::<usize>()` would be more idiomatic.

Reply:
Done.